pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go

```diff
@@ -585,6 +598,115 @@ func (grb *globalRoleBindingLifecycle) grantRestrictedAdminUserClusterPermission
 	return returnErr
 }
 
+// reconcileNamespacedRoleBindings ensures that RoleBindings exist for each namespace listed in NamespacedRules
+// from the associated GlobalRole
+func (grb *globalRoleBindingLifecycle) reconcileNamespacedRoleBindings(globalRoleBinding *v3.GlobalRoleBinding) error {
+	var returnError error
+	gr, err := grb.grLister.Get("", globalRoleBinding.GlobalRoleName)
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole %s: %w", globalRoleBinding.GlobalRoleName, err)
+	}
+
+	roleBindingUIDs := []types.UID{}
+
+	for ns := range gr.NamespacedRules {
+		_, err := grb.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+
+		roleName := gr.Name + "-" + ns
+		roleRef := v1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     roleName,
+		}
+
+		subject := rbac.GetGRBSubject(globalRoleBinding)
+		rbName := globalRoleBinding.Name + "-" + ns
+		roleBinding, err := grb.roleBindingLister.Get(ns, rbName)
+		if err == nil {
+			if !reflect.DeepEqual(roleRef, roleBinding.RoleRef) || roleBinding.Labels[grbOwnerLabel] != globalRoleBinding.Name {
+				// Since roleRef is immutable, we have to delete and recreate the RB
+				err = grb.roleBindings.DeleteNamespaced(roleBinding.Namespace, roleBinding.Name, &metav1.DeleteOptions{})
```

Comment:
More of a question: I understand that an update is not possible. Has this ever caused trouble (eg: in tests or for usres) where a user loses privileges shortly while a GRB is being reconciled?

For example:
1. Delete `RoleBinding`
2. Connectivity between Rancher & k8s fails momentarily
3. Reconnection
4. Finally create the `RoleBinding`

Maybe it's not too much of a problem in practice in 99% of the cases.

Reply:
Michael and I discussed something similar where a user could create an incorrect RoleBinding in the gap between delete and create, and we decided that it's a small enough chance that we are ok with it. Especially since any errors cause the reconcile to be re-queued.